### DISCUSSION TITLE
Does Oxigraph provide stable paging without ORDER BY?

### DISCUSSION THREAD

**Discussion by balhoff** [oxigraph/oxigraph#563] [Q&A] [answered]
I'd like to use Oxigraph as the backend for a [quad pattern fragments](https://linkeddatafragments.org/specification/quad-pattern-fragments/) server. There is a [JavaScript server implementation](https://github.com/LinkedDataFragments/Server.js) that allows pointing to a SPARQL endpoint as a data source. This involves sending simple one-pattern queries to the SPARQL endpoint, including a LIMIT and OFFSET so that the results can be paged through. But for performance reasons [they do not use ORDER BY](https://github.com/LinkedDataFragments/Server.js/blob/b8cc6e3c07d21371e11ea352924237cb48139855/packages/datasource-sparql/lib/datasources/SparqlDatasource.js#L126-L129). However this only works if the query results order is stable without ORDER BY—I'd like to avoid problems like this: https://github.com/LinkedDataFragments/Server.js/issues/24

Does Oxigraph provide any guarantees about stable ordering of results? Even better would be adding a built-in quad pattern fragments server, which I imagine could be a lot more efficient being implemented directly on Oxigraph internals; I could submit a feature request if there is any interest.

**Comment by Tpt** [comment, marked as answer]
Hi! Thank you for considering Oxigaph.

> Does Oxigraph provide any guarantees about stable ordering of results? 

Not explicitly but at the moment it does return results of a *single* triple pattern in a stable order. So `SELECT ?s ?p ?o WHERE { ?s ?p ?o } OFFSET X LIMIT Y` will be stable.

> Even better would be adding a built-in quad pattern fragments server, which I imagine could be a lot more efficient being implemented directly on Oxigraph internals; I could submit a feature request if there is any interest.

Yes, it would be great to have it! It's probably not a high priority feature but definitely a "nice to have".

**Reply by balhoff** [reply]
Thank you! I opened an issue: https://github.com/oxigraph/oxigraph/issues/565
